setuptools/msvc.py

```diff
@@ -173,39 +174,247 @@ def _msvc14_find_vc2015():
-            "-latest",
-            "-prerelease",
-            "-requires", "Microsoft.VisualStudio.Component.VC.Tools.x86.x64",
-            "-property", "installationPath",
-            "-products", "*",
-        ]).decode(encoding="mbcs", errors="strict").strip()
-    except (subprocess.CalledProcessError, OSError, UnicodeDecodeError):
-        return None, None
+        # create an instance of the ISetupConfiguration interface
+        setup_configuration = ctypes.c_void_p()
+        ole32.CoCreateInstance(
+            ctypes.byref(CLSID_SetupConfiguration),
+            None,
+            1, # CLSCTX_INPROC_SERVER
+            ctypes.byref(IID_ISetupConfiguration),
+            ctypes.byref(setup_configuration)
+        )
+
+        # setup the IUnknown::QueryInterface method
+        prototype = ctypes.WINFUNCTYPE(
+            ctypes.HRESULT,
+            ctypes.POINTER(GUID),
+            ctypes.POINTER(ctypes.c_void_p)
+        )
+        query_interface = prototype(0, 'QueryInterface')
+
+        # setup the IUnknown::Release method
+        prototype = ctypes.WINFUNCTYPE(ctypes.HRESULT)
+        release = prototype(2, 'Release')
+
+        # setup the ISetupConfiguration::EnumInstances method
+        prototype = ctypes.WINFUNCTYPE(
+            ctypes.HRESULT, ctypes.POINTER(ctypes.c_void_p)
+        )
+        enum_instances = prototype(3, 'EnumInstances')
+
+        enum_setup_instances = ctypes.c_void_p()
+        try:
+            # call ISetupConfiguration::EnumInstances and receive an
+            # instance of IEnumSetupInstances
+            enum_instances(
+                setup_configuration, ctypes.byref(enum_setup_instances)
+            )
+        finally:
+            # release the ISetupConfiguration instance
+            release(setup_configuration)
+
+        # setup the IEnumSetupInstances::Next method
+        prototype = ctypes.WINFUNCTYPE(
+            ctypes.HRESULT,
+            ctypes.c_ulong,
+            ctypes.POINTER(ctypes.c_void_p),
+            ctypes.POINTER(ctypes.c_ulong)
+        )
+        _next = prototype(3, 'Next')
+
+        # setup the ISetupInstance::GetInstallationPath method
+        prototype = ctypes.WINFUNCTYPE(
+            ctypes.HRESULT, ctypes.POINTER(ctypes.c_wchar_p)
+        )
+        get_installation_path = prototype(6, 'GetInstallationPath')
+
+        # setup the ISetupInstance2::GetPackages method
+        prototype = ctypes.WINFUNCTYPE(
+            ctypes.HRESULT, ctypes.POINTER(ctypes.c_void_p)
+        )
+        get_packages = prototype(12, 'GetPackages')
 
-    path = join(path, "VC", "Auxiliary", "Build")
-    if isdir(path):
-        return 15, path
+        setup_instance = ctypes.c_void_p()
+        setup_instance2 = ctypes.c_void_p()
+        installation_path = ctypes.c_wchar_p()
+        packages = ctypes.c_void_p()
+
+        try:
+            while _next(enum_setup_instances, 1, ctypes.byref(setup_instance), None) == 0: # S_OK
```

Comment:
I wasn't sure how do wrap this line to 79 characters without making it look like a mess. Any suggestions?